Migrate tests to use decorated new

mppf opened the pull request:
Relates to #9418 and #9087. This PR is a start at migrating test code away from un-decorated `new` since `new` is going to change behavior. This can be viewed 
as a follow-on to PR #9282.

Approach:
 * Fix problems with cast overloads for unmanaged in ChapelBase
 * Run testing with --warn-unstable
 * Use 'new borrowed', 'new owned' or 'new unmanaged' to resolve 
   --warn-unstable failures (this is done in a somewhat arbitrary manner 
   that migrates from undecorated new but without much thought)
 * Create issues for any cases that aren't easy

After this PR, expect ~700 failures with --warn-unstable, to be addressed in 
future PRs. Making examples/ or modules code use owned/shared/borrowed in the
best manner is also for future PRs (and that point is tracked in #9291 for the
tests modified here). Additionally there are quite a few issues representing
future work (just look under the delete-free epic #8623).

- [x] full local testing 

Reviewed by @benharsh, @ben-albrecht, @daviditen, and @vasslitvinov - thanks!

benharsh commented:
I have reviewed the following:
- [x] modules/internal/
- [x] spec/
- [x] test/arrays/
- [x] test/associative

Next up: classes/initializers/

benharsh commented on test/classes/initializers/compilerGenerated/generics/withParamTypeNoDef.chpl:
```diff
@@ -2,7 +2,7 @@ class Foo {
 }
 
-var foo = new Foo(true);
+var foo = new borrowed Foo(true);
```
'foo' is still deleted, is that acceptable?

vasslitvinov replied:
Good catch. `delete foo` should be a compiler error, right?

mppf replied:
Yeah, in the future, not yet. We have to migrate tests over before we can make that an error. But it's right that it shouldn't be deleted later.

benharsh commented:
The rest of classes/initializers/ looks good to me.

vasslitvinov commented on test/classes/diten/deep_nest.chpl:
```diff
@@ -32,37 +32,37 @@ class A {
 }
 
-var a = new A();
+var a = new borrowed A();
```
These uses of `borrowed` are cool!

vasslitvinov commented on test/classes/diten/returnMultipleTypes.chpl:
```diff
@@ -21,9 +21,9 @@ class C:B {
+    return new owned B();
   else
-    return new C();
+    return new owned C();
```
Also a cool use of `owned`.

daviditen commented on test/studies/ssca2/graphio/SSCA2_RMAT_graph_generator.chpl:
```diff
@@ -110,9 +110,9 @@ proc Gen_RMAT_graph ( a : real,
     // Random Numbers return in the range [0.0, 1.0)
 
     var Rand_Gen = if REPRODUCIBLE_PROBLEMS then
```
I don't think Rand_Gen is deleted.  Can it be borrowed?

mppf replied:
Can't be borrowed, because that runs into if-expr issues (that @benharsh is working on fixing).

mppf replied:
I made it owned and the test seems to pass, so we'll go with that for now

daviditen commented on test/release/examples/programs/linkedList.chpl:
```diff
@@ -150,10 +150,10 @@ class List {
   use Random;
 
-  var lst = new List(int);
+  var lst = new unmanaged List(int);
```
It doesn't look like `lst` is ever deleted.  Can it be borrowed instead?

ben-albrecht commented on test/types/type_variables/deitz/part1/test_clone_class1_old.chpl:
```diff
@@ -5,6 +5,6 @@ class foo {
 }
 
-var f : foo(2) = new foo(2);
+var f : foo(2) = new borrowed foo(2);
```
Is the `2`in `f: foo(2)` necessary here?

ben-albrecht commented on test/types/type_variables/deitz/part7/test_infer_class2.chpl:
```diff
@@ -3,10 +3,10 @@ class C {
 
 writeln(c);
 
-var d : C(int) = new C(int, 12);
```
Did this test previously leak?

mppf replied:
no, because it doesn't compile, it tests a compilation error

ben-albrecht commented on test/types/type_variables/deitz/test_query_expression_error.chpl:
```diff
@@ -3,9 +3,9 @@ class C {
 }
 
-var c1 = new C(2,1);
-var c2 = new C(6,2);
```
another leak fixed?

ben-albrecht commented:
My part:

- [x] modules/packages
- [x] test/trivial
- [x] test/types
- [ ] test/users
- [x] test/variables
- [x] test/visibility
- [x] tools/mason

mppf commented on test/classes/bradc/arrayInClass/arrayDomInClassRecord-illegal.chpl:
```diff
@@ -12,14 +12,14 @@ class C {
 writeln("myC is: ", myC);
 
-delete myC;
+
```
Those are there so I don't have to update the .good file

mppf commented on test/classes/bharshbarg/generic-forwarding.chpl:
```diff
@@ -8,14 +8,14 @@ class Lower {
 }
 
-var u = new Upper(int);
+var u = new unmanaged Upper(int);
```
The owned version doesn't actually work right now. I'll make an issue about it.

mppf replied:
it's issue #9601.

mppf commented on test/errhandling/ferguson/output-format.chpl:
```diff
@@ -9,7 +9,7 @@ class MyError : Error {
 }
 proc throwit() throws {
-  var e = new MyError("test error");
+  var e = new unmanaged MyError("test error");
```
The error handling (compiler-added) code assumes its responsible for deleting a thrown error. In the future though I hope it will generalize this stuff with `owned`. Right now `unmanaged` and `borrowed` are still close enough that you can delete a borrow - which is why it doesn't leak if you throw a borrow. But it should be an error.

mppf replied:
I made issue #9602 (mainly as a placeholder) to track that it shouldn't work in the future to throw a borrowed.

mppf commented on test/functions/bradc/retConstness/modifyReturnedClass.chpl:
```diff
@@ -1,13 +1,13 @@
 
 proc bar() {
-  var c2 = new C(2);
+  var c2 = new unmanaged C(2);
```
as you wish

mppf commented on test/library/standard/DataFrames/psahabu/AccessSeries.chpl:
```diff
@@ -3,7 +3,7 @@ use DataFrames;
 var I = ["A", "B", "C", "D", "E"];
 
-var letters = new TypedSeries(A, new TypedIndex(I));
+var letters = new unmanaged TypedSeries(A, new unmanaged TypedIndex(I));
```
I agree but I'd like @psahabu to take that on, in case there are non-trivial issues with how data frame classes refer to each other.

mppf commented on test/studies/590o/blerner/nilClassCheck.chpl:
```diff
@@ -7,7 +7,7 @@ const D = {1..10};
 
 for i in 1..5 {
-  A(i) = new C(i);
+  A(i) = new unmanaged C(i);
```
they couldn't be borrowed because they'd be deallocated at the end of the loop iteration. I think it's OK for them to be unmanaged in the context of this test.

mppf commented on test/studies/madness/aniruddha/madchap/mytests/par-refine/test_refine.chpl:
```diff
@@ -6,12 +6,14 @@ proc main() {
 
-    var fcn  : [1..4] AFcn = ((new Fn_Test1():AFcn),  (new Fn_Test2():AFcn),  (new Fn_Test3():AFcn), (new Fn_Unity():AFcn));
-    var dfcn : [1..4] AFcn = ((new Fn_dTest1():AFcn), (new Fn_dTest2():AFcn), (new Fn_dTest3():AFcn), (new Fn_dUnity():AFcn));
+    var fcn  : [1..4] unmanaged AFcn = ((unmanaged new Fn_Test1():unmanaged
```
thanks, not sure why it passed testing

mppf commented:
@daviditen, (and @vasslitvinov - fyi) I updated with your specific feedback. Regarding the general feedback:

> test/release/examples/benchmarks/shootout/chameneosredux.chpl

the array of unmanaged is a field and I can't currently convert it to owned without getting compilation errors. But, we can make the later new unmanaged & delete a new borrowed, so I did that.

> test/release/examples/benchmarks/ssca2/SSCA2_Modules/SSCA2_driver.chpl

I can't migrate them to borrowed because they are created in a loop.
I can't migrate them to owned right now because of compilation errors.


> test/release/examples/programs/linkedList.chpl

these can't be borrowed because they are fields that "own" the memory and you can't delete a borrowed.

mppf commented on modules/packages/TOML.chpl:
```diff
@@ -267,48 +267,55 @@ module TomlParser {
           while toStr.endsWith('"""') == false {
             toStr += " " + getToken(source);
           }
-          var mlString: Toml = toStr.strip('"""');
```
Actually it has to do with initialization is not the same as assignment. In this case, the TOML module supplies a `proc =` overload and expects it to be called in this case. It isn't necessarily called on initialization (but the compiler currently tries to do that, for better or worse, but it wasn't working with unmanaged).

mppf commented:
At this point I believe I've addressed all the feedback and I think I'm just waiting for @ben-albrecht to finish looking at test/users.

Thanks for your help, everybody!
